lib/modules/datasource/maven/util.ts

```diff
@@ -91,13 +93,62 @@ export async function downloadHttpProtocol(
       // istanbul ignore next
       logger.debug({ failedUrl }, 'Unsupported host');
     } else {
-      logger.info({ failedUrl, err }, 'Unknown error');
+      logger.info({ failedUrl, err }, 'Unknown HTTP download error');
     }
     return {};
   }
 }
 
-export async function checkHttpResource(
+function isS3NotFound(err: { name: string; message: string }): boolean {
+  return err.message === 'NotFound' || err.message === 'NoSuchKey';
+}
+
+export async function downloadS3Protocol(
+  pkgUrl: URL | string
+): Promise<string> {
+  logger.trace({ url: pkgUrl.toString() }, `Attempting to load S3 dependency`);
+  let body: string;
+  try {
+    const s3Url = parseS3Url(pkgUrl.toString());
+    if (s3Url === null) {
+      // istanbul ignore next
+      return '';
+    }
+    const response = await getS3Client().getObject(s3Url);
+    const stream = response.Body as Readable;
+    const buffers = await new Promise<Buffer>((resolve, reject) => {
+      const chunks: Buffer[] = [];
+      stream.on('data', (chunk) => chunks.push(chunk));
+      stream.once('end', () => resolve(Buffer.concat(chunks)));
+      stream.once('error', reject);
+    });
+    body = buffers.toString();
+    return body;
+  } catch (err) {
+    const failedUrl = pkgUrl.toString();
+    if (err.name === 'CredentialsProviderError') {
+      logger.debug(
+        { failedUrl },
+        'Dependency lookup authorization failed. Please correct AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY env vars'
```

Comment:
I'm not an expert in the cloud infrastructure, but can I ask on which side `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY` should be corrected? My guess it's on the cloud provider side, not Renovate. And if so, aren't those env var names specific to one particular vendor, i.e. AWS?

Reply:
Those two env vars are pretty common when working with AWS. You'd supply them in the runtime in which you run Renovate.

Reply:
Okay, hopefully users of any other S3-compatible storage would get it